File: Java/2370 Longest Ideal Subsequence.java
//  Problem: https://leetcode.com/problems/longest-ideal-subsequence

class Solution {
    public int longestIdealString(String s, int k) {
        // Let dp[c] be the maximum length of any ideal subsequence that ends with char c.
        int dp[] = new int[149], max = 0;

        // For each c in s, dp[c] is the maximum dp[i] within k of c plus one.
        for (char c : s.toCharArray()) {
            int localMax = 0;
            for (int j = c - k; j <= c + k; j++)
                localMax = Math.max(localMax, dp[j]);
            max = Math.max(max, dp[c] = ++localMax);
        }

        return max;
    }
}

File: C#/2370 Longest Ideal Subsequence.cs
//  Problem: https://leetcode.com/problems/longest-ideal-subsequence/

public class Solution {
    public int LongestIdealString(string s, int k) {
        var lastLength = new int[123];
        var S = s.ToCharArray();
        var max = 1;

        lastLength[S[S.Length - 1]] = 1;

        for (int i = S.Length - 2; i >= 0; i--) {
            var c = S[i];

            var low = Math.Max('a', c - k);
            var high = Math.Min('z', c + k);
            var m = 0;

            while (low <= high) {
                m = Math.Max(m, lastLength[low++]);
            }
            m++;

            lastLength[c] = m;

            max = Math.Max(max, m);
        }

        return max;
    }
}
